src/circuit_proof/mod.rs

```diff
@@ -0,0 +1,781 @@
+        let h = s_inv
+            .zip(util::exp_iter(y.invert()))
+            .map(|(s_i_inv, exp_y_inv)| -exp_y_inv * b * s_i_inv - Scalar::one());
+
+        // define parameters for t check
+        let delta = inner_product(&W_R_flatten_yinv, &W_L_flatten);
+        let powers_of_z: Vec<Scalar> = util::exp_iter(z).take(circuit.q).collect();
+        let z_c = z * inner_product(&powers_of_z, &circuit.c);
+        let W_V_flatten: Vec<Scalar> = matrix_flatten(circuit.W_V, z, circuit.m)?;
+        let V_multiplier = W_V_flatten.iter().map(|W_V_i| r * x * x * W_V_i);
+
+        let mega_check = RistrettoPoint::vartime_multiscalar_mul(
+            iter::once(x) // A_I
+                .chain(iter::once(x * x)) // A_O
+                .chain(iter::once(x)) // W_L_point
+                .chain(iter::once(x)) // W_R_point
+                .chain(iter::once(Scalar::one())) // W_O_point
+                .chain(iter::once(x * x * x)) // S
+                .chain(iter::once(w * (self.t_x - a * b) + r * (x * x * (delta + z_c) - self.t_x))) // B
+                .chain(iter::once(-self.e_blinding - r * self.t_x_blinding)) // B_blinding
+                .chain(g) // G
+                .chain(h) // H
+                .chain(x_sq.iter().cloned()) // ipp_proof.L_vec
+                .chain(x_inv_sq.iter().cloned()) // ipp_proof.R_vec
+                .chain(V_multiplier) // V
+                .chain(iter::once(r * x)) // T_1
+                .chain(iter::once(r * x * x * x)) // T_3
+                .chain(iter::once(r * x * x * x * x)) // T_4
+                .chain(iter::once(r * x * x * x * x * x)) // T_5
+                .chain(iter::once(r * x * x * x * x * x * x)), // T_6
+            iter::once(&self.A_I)
+                .chain(iter::once(&self.A_O))
+                .chain(iter::once(&W_L_point))
+                .chain(iter::once(&W_R_point))
+                .chain(iter::once(&W_O_point))
+                .chain(iter::once(&self.S))
+                .chain(iter::once(&gen.pedersen_generators.B))
+                .chain(iter::once(&gen.pedersen_generators.B_blinding))
+                .chain(gen.G.iter())
+                .chain(gen.H.iter())
+                .chain(self.ipp_proof.L_vec.iter())
+                .chain(self.ipp_proof.R_vec.iter())
+                .chain(V.iter())
+                .chain(iter::once(&self.T_1))
+                .chain(iter::once(&self.T_3))
+                .chain(iter::once(&self.T_4))
+                .chain(iter::once(&self.T_5))
+                .chain(iter::once(&self.T_6)),
+        );
+
+        if !mega_check.is_identity() {
+            return Err("Circuit did not verify correctly.");
+        }
+
+        Ok(())
+    }
+}
+
+// Computes z * z^Q * W, where W is a qxn matrix and z is a scalar.
+// Input: Qxn matrix of scalars and scalar z
+// Output: length n vector of Scalars
+pub fn matrix_flatten(
+    W: Vec<Vec<Scalar>>,
+    z: Scalar,
+    output_dim: usize,
```

Comment:
What's the role of the `output_dim` parameter? Since the matrix has a known size, why not use that size instead of making the caller pass a parameter in and check consistency?

Reply:
The reason we need the `output_dim` parameter is if the matrix is size `qxn` where `q=0` and `n>0` (such as in the `mul_circuit_2` test case). In this case, you could not figure out what the output vector's size should be (since you can't figure out `n` by looking at the matrix), but the output of the `matrix_flatten` function should still be a length `n` vector. 

It's possible to do a check higher up for if `n=0`, before calling `matrix_flatten`, but I thought it would be cleaner to pass in `n` as `output_dim` and have the `matrix_flatten` function take care of those checks internally.